File: Sources/UsageWidget/Providers/Copilot/CopilotProvider.swift
import Foundation
import AppKit

/// GitHub Copilot provider: device-flow OAuth sign-in, token refresh, and monthly-quota fetch.
actor CopilotProvider: UsageProvider {
    nonisolated let id = ProviderID.copilot
    nonisolated let displayName = "Copilot"
    nonisolated let iconSystemName = "cpu"
    nonisolated let accentHex = "#A78BFA"
    nonisolated let defaultPollInterval: TimeInterval = 600  // monthly quota moves slowly
    nonisolated let minimumPollInterval: TimeInterval = 120

    private let tokens: TokenStore
    private let oauth = CopilotOAuthClient()
    private let fetcher = CopilotUsageFetcher()

    init(tokens: TokenStore) {
        self.tokens = tokens
    }

    func authState() async -> ProviderAuthState {
        await tokens.token(for: id) != nil ? .signedIn : .signedOut
    }

    func startSignIn() async throws -> SignInContinuation {
        let device = try await oauth.requestDeviceCode()
        let url = URL(string: device.verificationURI) ?? URL(string: "https://github.com/login/device")!
        await openInBrowser(url)

        let oauth = self.oauth
        let tokens = self.tokens
        let id = self.id
        return .deviceCode(
            userCode: device.userCode,
            verificationURL: url,
            instructions: "Enter this code at github.com/login/device to connect Copilot."
        ) {
            let token = try await oauth.pollForToken(deviceCode: device.deviceCode, interval: device.interval, expiresIn: device.expiresIn)
            await tokens.save(token, for: id)
        }
    }

    func signOut() async {
        await tokens.clear(id)
    }

    func fetchUsage() async throws -> ProviderUsage {
        guard var token = await tokens.token(for: id) else { throw ProviderError.notSignedIn }
        if token.needsRefresh(), token.refreshToken != nil {
            token = (try? await refreshAndStore(token)) ?? token
        }
        do {
            return try await fetcher.fetch(accessToken: token.accessToken)
        } catch ProviderError.unauthorized {
            if token.refreshToken != nil {
                let refreshed = try await refreshAndStore(token)
                return try await fetcher.fetch(accessToken: refreshed.accessToken)
            }
            await tokens.clear(id)  // non-expiring token was revoked → require re-auth
            throw ProviderError.unauthorized
        }
    }

    private func refreshAndStore(_ token: OAuthToken) async throws -> OAuthToken {
        do {
            let refreshed = try await oauth.refresh(token)
            await tokens.save(refreshed, for: id)
            return refreshed
        } catch {
            await tokens.clear(id)
            throw ProviderError.unauthorized
        }
    }

    @MainActor
    private func openInBrowser(_ url: URL) {
        NSWorkspace.shared.open(url)
    }
}

File: windows/src/Core/Providers/CopilotProvider.cs
using LLMUsageWidget.Core.Auth;
using LLMUsageWidget.Core.Domain;

namespace LLMUsageWidget.Core.Providers;

/// <summary>GitHub Copilot provider: device-flow OAuth, token refresh, and monthly-quota fetch.</summary>
public sealed class CopilotProvider : IUsageProvider
{
    public ProviderId Id => ProviderId.Copilot;
    public string DisplayName => "Copilot";
    public string AccentHex => "#8957E5";
    public TimeSpan DefaultPollInterval => TimeSpan.FromSeconds(600);  // monthly quota moves slowly
    public TimeSpan MinimumPollInterval => TimeSpan.FromSeconds(120);

    private readonly ITokenStore _tokens;
    private readonly Action<Uri> _openBrowser;
    private readonly CopilotOAuthClient _oauth = new();
    private readonly CopilotUsageFetcher _fetcher = new();

    public CopilotProvider(ITokenStore tokens, Action<Uri> openBrowser)
    {
        _tokens = tokens;
        _openBrowser = openBrowser;
    }

    public Task<ProviderAuthState> AuthStateAsync() =>
        Task.FromResult(_tokens.Get(Id) is not null ? ProviderAuthState.SignedIn : ProviderAuthState.SignedOut);

    public async Task<SignInFlow> StartSignInAsync()
    {
        var device = await _oauth.RequestDeviceCodeAsync();
        var url = Uri.TryCreate(device.VerificationUri, UriKind.Absolute, out var u) ? u : new Uri("https://github.com/login/device");
        _openBrowser(url);
        Func<Task> poll = async () =>
        {
            var token = await _oauth.PollForTokenAsync(device.DeviceCode, device.Interval, device.ExpiresIn);
            _tokens.Save(Id, token);
        };
        return new SignInFlow.DeviceCode(device.UserCode, url,
            "Enter this code at github.com/login/device to connect Copilot.", poll);
    }

    public Task SignOutAsync()
    {
        _tokens.Clear(Id);
        return Task.CompletedTask;
    }

    public async Task<ProviderUsage> FetchUsageAsync()
    {
        var token = _tokens.Get(Id) ?? throw ProviderException.NotSignedIn();
        if (token.NeedsRefresh() && token.RefreshToken is not null)
        {
            try { token = await RefreshAndStoreAsync(token); }
            catch (ProviderException) { /* keep current token; the fetch below will decide */ }
        }
        try
        {
            return await _fetcher.FetchAsync(token.AccessToken);
        }
        catch (ProviderException ex) when (ex.Kind == ProviderErrorKind.Unauthorized)
        {
            if (token.RefreshToken is not null)
            {
                var refreshed = await RefreshAndStoreAsync(token);
                return await _fetcher.FetchAsync(refreshed.AccessToken);
            }
            _tokens.Clear(Id);  // non-expiring token was revoked → require re-auth
            throw;
        }
    }

    private async Task<OAuthToken> RefreshAndStoreAsync(OAuthToken token)
    {
        try
        {
            var refreshed = await _oauth.RefreshAsync(token);
            _tokens.Save(Id, refreshed);
            return refreshed;
        }
        catch
        {
            _tokens.Clear(Id);
            throw ProviderException.Unauthorized();
        }
    }
}
